Adding MaterialRealCMMAux and some other changes.

cbolisetti opened the pull request:
Makes the following changes/additions: 

- Add the MaterialRealCMMAux, tests, and doco. 
- Changes to linear spring material property variables to convert them to ColumnMajorMatrix instead, so that they are consistent with isolator materials. 
- Other miscellaneous changes and addressing a doco bug. 

Closes #374 #356 and #326

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/756599/) on da1839b wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/mastodon/docs/PRs/376/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/mastodon/docs/PRs/376/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 789f12261b7c00af7b0ae8fd5d45a2d0f4b88950`

cbolisetti commented:
@somu15 can you please review this?

somu15 commented on doc/content/source/auxkernels/MaterialRealCMMAux.md:
```diff
@@ -0,0 +1,32 @@
+a material property that is of type, ColumnMajorMatrix (CMM). In MASTODON, a
+common use case is accessing the deformation or force CMMs from link elements
+when the [`LinearSpring`](source/materials/LinearSpring.md) material or any of
+the seismic isolator materials (e.g., [`ComputeFPIsolatorElasticity`](source/materials/ComputeFPIsolatorElasticity.md)) is used. This `AuxKernel` and the corresponding
```
Does this mean we can directly get the isolator deformations without having to do a difference between top and bottom node displacements? If so, can you please state this explicitly so that the benefits of this auxkernel are clearer?

somu15 commented on src/auxkernels/MaterialRealCMMAux.C:
```diff
@@ -0,0 +1,45 @@
+  InputParameters params = MaterialAuxBase<>::validParams();
+  params.addClassDescription(
+      "Populate an auxiliary variable with an entry from a ColumnMajorMatrix material property.");
+  params.addParam<unsigned int>("row", 0, "The row component to consider for this kernel");
```
How will the user know which row to use here?

somu15 commented on src/auxkernels/MaterialRealCMMAux.C:
```diff
@@ -0,0 +1,45 @@
+  params.addClassDescription(
+      "Populate an auxiliary variable with an entry from a ColumnMajorMatrix material property.");
+  params.addParam<unsigned int>("row", 0, "The row component to consider for this kernel");
+  params.addParam<unsigned int>("column", 0, "The column component to consider for this kernel");
```
Similar to previous, how to know which column to use?

somu15 commented on src/kernels/StressDivergenceDamper.C:
```diff
@@ -33,8 +33,8 @@ StressDivergenceDamper::StressDivergenceDamper(const InputParameters & parameter
     _disp_var(_ndisp),
-    _Fg(getMaterialPropertyByName<ColumnMajorMatrix>("global_forces")),
-    _Kg(getMaterialPropertyByName<ColumnMajorMatrix>("global_stiffness_matrix"))
+    _Fg(getMaterialPropertyByName<ColumnMajorMatrix>("fvd_global_forces")),
```
You are changing these names so that there is consistency across all line type materials? Like Spring, Isolator, etc.?

somu15 commented on src/materials/ComputeLRIsolatorElasticity.C:
```diff
@@ -107,31 +109,33 @@ ComputeLRIsolatorElasticity::ComputeLRIsolatorElasticity(const InputParameters &
-    _total_gl(getMaterialPropertyByName<ColumnMajorMatrix>("total_global_to_local_transformation")),
-    _total_lb(getMaterialPropertyByName<ColumnMajorMatrix>("total_local_to_basic_transformation")),
-    _length(getMaterialPropertyByName<Real>("initial_isolator_length")),
+    _basic_def(getMaterialPropertyByName<ColumnMajorMatrix>("si_deformations")),
```
Can you please explain the reason for changing these names?

somu15 reviewed (commented):
@cbolisetti I did an initial review. Let me know if you have any questions.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/758145/) on 2199c99 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/mastodon/docs/PRs/376/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/mastodon/docs/PRs/376/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format a9ccdaf6572ceb7b112cd8e5c95d699ca19aa57d`

cbolisetti commented:
@somu15 thanks for the review. I was going to remove the commits on changes in material properties, but I forgot -- my bad. Those were supposed to address #356. I am not yet sure what the solution for that is, but I will leave it for now. Please check out the commits now. I added some more detail in the auxkernel doco. Note that this is similar to the RankTwoTensor auxkernel, in which the user needs to know which component of the tensor corresponds to which stress component, etc. I added some language about this in the end.

somu15 commented on doc/content/source/auxkernels/MaterialRealCMMAux.md:
```diff
@@ -0,0 +1,19 @@
+
+!listing test/tests/auxkernels/materialrealcmm/spring_static.i block=AuxKernels
+
+In this input, the AuxKernel takes the element (0, 0) from the CMM material property named, "global_forces" in the linear spring material. (See LinearSpring.C for other material properties.) In this example, the linear spring material is only defined in block '0' and therefore this AuxKernel and AuxVariable are also restricted to this block. The element (0, 0) of the material property "global_for ...
```
This information is very helpful to the users.

somu15 reviewed (commented):
@cbolisetti This looks good. One comment is, if the user requests a row/column value outside the validity range, do you want to throw a moose error in the auxkernel? Or else, what is your thought on this?

cbolisetti commented:
> @cbolisetti This looks good. One comment is, if the user requests a row/column value outside the validity range, do you want to throw a moose error in the auxkernel? Or else, what is your thought on this?

@somu15 yes, the AuxKernel will throw an error when that happens. See lines 42 and 43 of MaterialRealCMMAux.C

somu15 commented:
> > @cbolisetti This looks good. One comment is, if the user requests a row/column value outside the validity range, do you want to throw a moose error in the auxkernel? Or else, what is your thought on this?
> 
> @somu15 yes, the AuxKernel will throw an error when that happens. See lines 42 and 43 of MaterialRealCMMAux.C

Sounds good! Thanks for the clarification.
